lib/bookmark_query.rb

```diff
@@ -35,13 +35,22 @@ def list_all
     pms = Topic.private_messages_for_user(@user)
     results = results.merge(topics.or(pms))
 
-    results = results.merge(Post.secured(@guardian))
+    results = results.where(
+      "posts.post_type IN (?) OR bookmarks.post_id IS NULL", Topic.visible_post_types(@guardian&.user)
+    )
 
     if @params[:q].present?
       term = @params[:q]
       bookmark_ts_query = Search.ts_query(term: term)
       results = results
-        .joins("LEFT JOIN post_search_data ON post_search_data.post_id = bookmarks.post_id")
+        .joins(<<~SQL)
+          LEFT JOIN post_search_data ON post_search_data.post_id = CASE
```

Comment:
This is ugly but I am not really sure how else to do this...

Reply:
Should we be searching against the `post_search_data` for topic bookmarks? I think topic level bookmarks should be searching against `topic_search_data` instead.